### ISSUE TITLE
CORS helpers

### ISSUE THREAD

**Issue by jyecusch** [nitrictech/nitric#520]
While CORS can currently be handled using middleware and by registering OPTIONS routes, we'd like to build this support into the framework to make it more convinient and ideally handle the responses as high up the infra stack as possible. For example, providing support for static responses to OPTIONS requests at the API Gateway or Nitric Server layers to reduce latency and hosting costs.

This support will be needed in all language SDKs.

**Comment by davemooreuws** [comment]
Research of cloud provider API gateway pulumi configs:

AWS:
Seems pretty easy, just put them in apigatewayv2.NewApi
- allowCredentials
- allowHeaders
- allowMethods
- allowOrigins
- exposeHeaders
- maxAge

Azure:
Need to create a [cors policy](https://learn.microsoft.com/en-us/azure/api-management/cors-policy) use apimanagement.NewApiOperationPolicy

- allowCredentials
- allowHeaders
- allowMethods
- allowOrigins
- exposeHeaders
- maxAge

GCP:
The worst of the bunch it seems, they suggest using ESP or doing it per function.
References:
https://stackoverflow.com/questions/64281334/cors-errors-when-trying-to-fetch-from-new-google-cloud-api-gateway/65738206#65738206
https://www.googlecloudcommunity.com/gc/Serverless/CORS-issue-with-API-gateway/m-p/619097

TLDR

AWS and Azure seem fine, however considering GCPs lack of support (and lack of a roadmap) Like we have suggested, we could use a load balancer or just handle it on our side, but then if we do this should we just do it consistently across any new provider or pick best prac per cloud?

**Comment by davemooreuws** [comment]
Another item for discussion. Should we also include bucket CORS configuration in this ticket?

**Comment by davemooreuws** [comment]
Started work on proto contracts / integration with AWS.

**Comment by raksiv** [comment]
Is the Bucket CORS suggestion for people who want to access the bucket from a resource other than our API?

**Comment by davemooreuws** [comment]
> Is the Bucket CORS suggestion for people who want to access the bucket from a resource other than our API?

Yes, but should be a new ticket if we decide to do it

**Comment by davemooreuws** [comment]
This has been completed and is ready for review (in this order):
- https://github.com/nitrictech/nitric/pull/523
Then
https://github.com/nitrictech/cli/pull/595
https://github.com/nitrictech/node-sdk/pull/206

**Comment by davemooreuws** [comment]
Reviving this feature for v1.

Example in ts:

```ts
// no cors by default
const helloApi = api("main");

// with custom settings
const helloApi = api("main", {
  cors: {
    allowOrigins: ["https://cors-test.codehappy.dev"],
    allowCredentials: true,
    allowMethods: ["GET"],
  },
});

// with default cors settings
const helloApi = api("public", {
  cors: true,
});
```

Implementations:
- AWS: via native apigateway
- Azure: via native apimanagement
- GCP: custom implementation in the nitric layer
- Local: custom implementation in the CLI

Old pre v1 reference work to use:
- https://github.com/nitrictech/nitric/pull/539
- https://github.com/nitrictech/cli/pull/595
- https://github.com/nitrictech/node-sdk/pull/206

**Comment by jyecusch** [comment]
No longer needed in proposed v2 design
